### PR TITLE
add n hits to rate limit request

### PR DESCRIPTION
**Pull request by junr03**
@lyft/network-team.

Limit stats (over limit, and total hits) were moved into the cache as well.

### CODE REVIEW

File: src/redis/cache_impl.go [hunk 1]
```diff
@@ -99,20 +117,47 @@ func (this *rateLimitCacheImpl) DoLimit(
 		current := uint32(conn.PipeResponse().Int())
 		conn.PipeResponse() // Pop off EXPIRE response and check for error.
 
+		limit := limits[i].Limit.RequestsPerUnit
+
+		// previous is the the value of the limit before adding addNHits.
+		previous := current - addNHits
```
Comment (mattklein123) [inline src/redis/cache_impl.go]: s/previous/limit_value_before_increment

File: src/redis/cache_impl.go [hunk 2]
```diff
@@ -99,20 +117,47 @@ func (this *rateLimitCacheImpl) DoLimit(
...
+		if current > limit {
 			responseDescriptorStatuses[i] =
 				&pb.RateLimitResponse_DescriptorStatus{pb.RateLimitResponse_OVER_LIMIT,
 					limits[i].Limit, 0}
+
+			// Increase over limit statistics. Because we support += behavior for increasing the limit, we need to
+			// asses if the entire addNHits were over the limit. That is if the limit's value before adding the
+			// N hits was over the limit, then all the N hits were over limit.
+			// Otherwise, only the difference between the current value and the limit were over limit hits.
+			if previous >= limit {
```
Comment (mattklein123) [inline src/redis/cache_impl.go]: I'm sorry but my brain hurts trying to make sense of the logic/math here. Can you try renaming variables to be more clear per some other comments and then I can look again.

File: proto/ratelimit/ratelimit.proto [hunk 3]
```diff
@@ -26,6 +26,9 @@ message RateLimitRequest {
   // processed by the service (see below). If any of the descriptors are over limit, the entire
   // request is considered to be over limit.
   repeated RateLimitDescriptor descriptors = 2;
+  // Rate limit requests can optionally specify the number of hits a request adds to the matched limit. If the
+  // value is not set in the message, a request increases the matched limit by 1.
+  uint32 add_n_hits = 3;
```
Comment (junr03) [inline proto/ratelimit/ratelimit.proto]: changed.

File: src/redis/cache_impl.go (hunk 2 above)
Comment (junr03) [inline src/redis/cache_impl.go]: changed some names and added more comments to make this more clear.

File: src/redis/cache_impl.go [hunk 4]
```diff
@@ -58,6 +59,13 @@ func (this *rateLimitCacheImpl) generateCacheKey(
 	return cacheKey
 }
 
+func max(a uint32, b uint32) uint32 {
```
Comment (junr03) [inline src/redis/cache_impl.go]: There is no other function like that. I could move this to an utility package, is it worth it if it is not used anywhere else?

Comment (mattklein123) [inline src/redis/cache_impl.go]: Sorry maybe I did this somewhere else. I can't remember now.

File: src/redis/cache_impl.go [hunk 5]
```diff
@@ -67,13 +75,23 @@ func (this *rateLimitCacheImpl) DoLimit(
 	conn := this.pool.Get()
 	defer this.pool.Put(conn)
 
+	var addNHits uint32 = request.AddNHits
+	if addNHits == 0 {
```
Comment (junr03) [inline src/redis/cache_impl.go]: this got collapsed without me noticing, adding comment.

File: src/redis/cache_impl.go [hunk 6]
```diff
@@ -96,23 +112,52 @@ func (this *rateLimitCacheImpl) DoLimit(
...
+		nearLimitThreshold := uint32(math.Floor(float64(float32(overLimitThreshold) * config.NearLimitRatio)))
+
+		logger.Debugf("cache key: %s current: %d", cacheKey, limitAfterIncrease)
+		if limitAfterIncrease > overLimitThreshold {
 			responseDescriptorStatuses[i] =
 				&pb.RateLimitResponse_DescriptorStatus{pb.RateLimitResponse_OVER_LIMIT,
 					limits[i].Limit, 0}
+
+			// Increase over limit statistics. Because we support += behavior for increasing the limit, we need to
+			// asses if the entire addend hits were over the limit. That is, if the limit's value before adding the
```
Comment (mattklein123) [inline src/redis/cache_impl.go]: assess

File: proto/ratelimit/ratelimit.proto [hunk 7]
```diff
@@ -26,6 +26,9 @@ message RateLimitRequest {
   // processed by the service (see below). If any of the descriptors are over limit, the entire
   // request is considered to be over limit.
   repeated RateLimitDescriptor descriptors = 2;
+  // Rate limit requests can optionally specify the number of hits a request adds to the matched limit. If the
+  // value is not set in the message, a request increases the matched limit by 1.
+  uint32 addend = 3;
```
Comment (ccaraman) [inline proto/ratelimit/ratelimit.proto]: nit: can you make this var nam a bit more intuitive?

File: test/redis/cache_impl_test.go [hunk 8]
```diff
@@ -154,5 +168,128 @@ func TestRedis(t *testing.T) {
 		[]*pb.RateLimitResponse_DescriptorStatus{
 			{pb.RateLimitResponse_OVER_LIMIT, limits[0].Limit, 0}},
 		cache.DoLimit(nil, request, limits))
+	assert.Equal(uint64(3), limits[0].Stats.TotalHits.Value())
+	assert.Equal(uint64(1), limits[0].Stats.OverLimit.Value())
+	assert.Equal(uint64(1), limits[0].Stats.NearLimit.Value())
+
+	// Now test addNHits that is greater than 1
```
Comment (mattklein123) [inline test/redis/cache_impl_test.go]: addend

**Pull request merged**